core/src/main/java/org/apache/iceberg/io/ClusteredWriter.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.Set;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeSet;
+
+/**
+ * A writer capable of writing to multiple specs and partitions ensuring the incoming records are properly clustered.
+ */
+public abstract class ClusteredWriter<T, R> implements PartitionAwareWriter<T, R> {
```

Comment:
Naming suggestions are welcome.

Reply:
I like "clustered" since that's the assumption.

Reply:
The naming has a slight conflict with the "sorted" position delete writer. That writer keeps deletes in memory and sorts them prior to writing them out. That's for the CDC use case where position deletes may be in any order, rather than the `MERGE` use case where we should be able to ask the engine to produce deletes in the expected file/pos order.

Reply:
Yeah, any ideas here? The main distinction is that `ClusteredWriter` writes to multiple specs/partitions and just checks the data is clustered. That writer actually sorts position deletes and writes only to a single spec/partition.